x-pack/libbeat/common/cloudfoundry/cache.go

```diff
@@ -0,0 +1,87 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package cloudfoundry
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/client-go/tools/cache"
+
+	cfclient "github.com/cloudfoundry-community/go-cfclient"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// cfClient interface is provided so unit tests can mock the actual client.
+type cfClient interface {
+	// GetAppByGuid returns an application information from its Guid.
+	GetAppByGuid(guid string) (cfclient.App, error)
+}
+
+// clientCacheWrap wraps the cloudfoundry client to add a cache in front of GetAppByGuid.
+type clientCacheWrap struct {
+	store  cache.Store
+	client cfClient
+	log    *logp.Logger
+	ttl    time.Duration
+}
+
+// newClientCacheWrap creates a new cache for application data.
+func newClientCacheWrap(client cfClient, ttl time.Duration, log *logp.Logger) *clientCacheWrap {
+	return &clientCacheWrap{
+		store:  cache.NewTTLStore(cacheKeyFunc, ttl),
+		client: client,
+		ttl:    ttl,
+		log:    log,
+	}
+}
+
+// appCached wraps an App structure adding a retrieval time
+type appCached struct {
+	app cfclient.App
+	ttl time.Time
+}
+
+// fetchApp uses the cfClient to retrieve an App entity and
+// stores it in the internal cache
+func (c *clientCacheWrap) fetchAppByGuid(guid string) (*cfclient.App, error) {
+	app, err := c.client.GetAppByGuid(guid)
+	if err != nil {
+		return nil, err
+	}
+	ttl := time.Now().Add(c.ttl)
+	c.store.Add(&appCached{app, ttl})
```

Comment:
We are never deleting elements from this cache, this could be a leak.

In case it can be useful here, we also have a cache implementation in `libbeat/common` that can optionally start a goroutine (`StartJanitor` and `StopJanitor` for finalization) for cleanup of expired entries.

Reply:
Looking at the cache in `libbeat/common` seems it only does eviction based on access time so this would do what I was trying to prevent in the other cache code above.

Maybe I should switch to this cache, but add a new way for it to operate? Because looked at the cache I am using from Kubernetes tools and it doesn't remove other expired items when you access any item, so as soon as an App is no longer referenced it will just live in the cache forever.

Reply:
I added a way to use the `libbeat/common` cache.

Reply:
Thanks!